### PR TITLE
Target and Count encodings for categorical features - Iteration 1

### PR DESCRIPTION
**Pull request by tongwu-msft**
Split PR: [Target and Count encodings for categorical features](https://github.com/microsoft/LightGBM/pull/3234) to multiple part. 

## Implementation
1. CategoryFeatureEncoder: abstract class for encoders.
2. CategoryFeatureEncoderManager: encode value for aggregated multiple encoders.
3. CategoryFeatureTargetInformationCollector: collect data information before training.

### CODE REVIEW

File: src/feature_engineering/category_feature_target_encoder.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+// property name keys
+const std::string count_information_key = "count_information";
+const std::string label_information_key = "label_information";
+const std::string category_key = "cat";
+const std::string value_key = "value";
+const std::string count_prior_key = "prior";
+const std::string count_prior_weight_key = "prior_weight";
+
+namespace LightGBM {
+  double CategoryFeatureTargetEncoder::Encode(double feature_value) {
```
Comment (shiyu1994) [inline src/feature_engineering/category_feature_target_encoder.cpp:17]: Is the fold ID considered for encoding of training dataset?

Comment (shiyu1994) [reply src/feature_engineering/category_feature_target_encoder.cpp:17]: Ok. I see that the encoder is per-fold. So there's no question.

**Comment by shiyu1994** [comment]
@tongwu-msft Thank you for working on this. The design overall looks good to me. Can we move on to integrating with the dataset classes?

**Comment by tongwu-msft** [comment]
> @tongwu-msft Thank you for working on this. The design overall looks good to me. Can we move on to integrating with the dataset classes?

Yes, let me split the big PR into multiple iterations, this is it1 and we can do it parallel.

**Comment by jameslamb** [comment]
I'm very confused. After 7 months of inactivity, this PR was closed without any comment...and it was supposed to replace #3234 which has not received a commit in about 8 months.

@shiyu1994 @tongwu-msft @guolinke what is happening with the large changes to LightGBM's handling of categorical values? Is someone working on this and should it still be considered a requirement for v4.0.0 (#5153)?

cc @jmoralez @StrikerRUS @btrotta

**Comment by shiyu1994** [comment]
I can pick this up. But I'll first focus on the CUDA parts before v4.0.0. Another option would be extracting the categorical feature handling into a separate module for LightGBM dataset preprocessing (something like https://github.com/microsoft/LightGBM-transform). So that we can focus on the core training algorithm of GBDT in our codebase.

**Pull request closed without merging**